core/src/main/java/org/elasticsearch/common/collect/Iterators.java

```diff
@@ -0,0 +1,121 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.collect;
+
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+public class Iterators {
+    public static <T> Iterator<T> unmodifiableIterator(Iterator<T> iterator) {
+        return iterator instanceof UnmodifiableIterator ? iterator : new UnmodifiableIterator<>(iterator);
+    }
+
+    static class UnmodifiableIterator<T> implements Iterator<T> {
+        private final Iterator<T> iterator;
+
+        public UnmodifiableIterator(Iterator<T> iterator) {
+            if (iterator == null) {
+                throw new NullPointerException("iterator");
+            }
+            this.iterator = iterator;
+        }
+
+        @Override
+        public boolean hasNext() {
+            return iterator.hasNext();
+        }
+
+        @Override
+        public T next() {
+            return iterator.next();
+        }
+    }
+
+    public static <T> T[] toArray(Iterator<? extends T> iterator, Class<T> type) {
```

Comment:
Similarly it looks like usage of this method always looks like `Iterators.toArray(stream.iterator())` so maybe we could directly use Stream.toArray?